src/number_theory/pell.lean

```diff
@@ -137,4 +137,119 @@ end
 
 end existence
 
+/-!
+### Group structure of the solution set
+
+We define a structure of a commutative multiplicative group with distributive negation
+on the set of all solutions to the Pell equation `x^2 - d*^2 = 1`.
+
+The type of such solutions is `pell.solution₁ d`. It contains integers `x` and `y` and
+a proof that `(x, y)` is indeed a solution.
+
+The multiplication is given by `(x, y) * (x', y') = (x*y' + d*y*y', x*y' + y*x')`
+(this is obtained by mapping `(x, y)` to `x + y*√d` and multiplying the results).
+-/
+
+-- We use `solution₁ d` to allow for a more general structure `solution d m` that
+-- encodes solutions to `x^2 - d*y^2 = m` to be added later.
+
+/-- `pell.solution₁ d` is the type of solutions to the Pell equation `x^2 - d*y^2 = 1`.
+This is a structure containing two integers `x` and `y` and a proof `rel` that the equation holds.
+-/
+structure solution₁ (d : ℤ) := (x : ℤ) (y : ℤ) (rel : x ^ 2 - d * y ^ 2 = 1)
+
+namespace solution₁
+
+instance (d : ℤ) : inhabited (solution₁ d) := ⟨{x := 1, y := 0, rel := by simp}⟩
+
+variables {d : ℤ}
+
+/-- An alternative form of the relation, suitable for rewriting `x^2`. -/
+lemma rel_x (a : solution₁ d) : a.x ^ 2 = 1 + d * a.y ^ 2 := by {rw ← a.rel, ring}
+
+/-- An alternative form of the relation, suitable for rewriting `d * y^2`. -/
+lemma rel_y (a : solution₁ d) : d * a.y ^ 2 = a.x ^ 2 - 1 := by {rw ← a.rel, ring}
+
+/-- Two solutions are equal if their `x` and `y` components are equal. -/
+@[ext]
+lemma ext {a b : solution₁ d} (hx : a.x = b.x) (hy : a.y = b.y) : a = b :=
+begin
+  cases a,
+  cases b,
+  dsimp at hx hy,
+  simp only [hx, hy, eq_self_iff_true, and_self],
+end
+
+/-- We use `1` to denote the trivial solution `(1, 0)`. -/
+def one : solution₁ d :=
```

Comment:
Can you write this as a `has_one` instance to avoid needing to name it? Same below.

Reply:
Yes. (Same with `mul`, `inv`, `neg`.)